controller/integration/test/finalizer/finalizer_test.go

```diff
@@ -0,0 +1,331 @@
+// +build k8srequired
+
+package parallel
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+	"testing"
+	"time"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	"github.com/giantswarm/backoff"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/operatorkit/controller/integration/testresource"
+	"github.com/giantswarm/operatorkit/controller/integration/wrapper/drainerconfig"
+	"github.com/giantswarm/operatorkit/resource"
+)
+
+const (
+	// controllerName is the same for all three controllers in order to simulate
+	// multiple VOO deployments of the same operator.
+	controllerName = "test-controller"
+	// finalizer is the only finalizer shared by the three test controllers.
+	finalizer    = "operatorkit.giantswarm.io/test-controller"
+	objName      = "test-obj"
+	objNamespace = "integration-finalizer-test"
+)
+
+// Test_Controller_Integration_Finalizer is an integration test to check that
+// multiple controllers do not interfere with finalizer management where they
+// should not act. This is to prevent situations in which multiple operators
+// reconcile the same runtime object in different versions. Then finalizer
+// management should only be done by controllers that express to handle the
+// reconciliation for an object.
+func Test_Controller_Integration_Finalizer(t *testing.T) {
```

Comment:
I hope the comments here and in the test are useful and explain what this all is about.